lib/aliquot/payment.rb

```diff
@@ -0,0 +1,129 @@
+module Aliquot
+  ##
+  # A Payment represents a single payment using Google Pay.
+  # It is used to verify/decrypt the supplied token by using the shared secret,
+  # thus avoiding having knowledge of merchant primary keys.
+  class Payment
+    ##
+    # Parameters:
+    # token_string::  Google Pay token (JSON string)
+    # shared_secret:: Base64 encoded shared secret
+    # merchant_id::   Google Pay merchant ID ("merchant:<SOMETHING>")
+    # logger::        The logger to use. Default: Logger.new($stdout)
+    # signing_keys::  Formatted list of signing keys used to sign token contents.
+    #                 Otherwise a thread continuously updating google signing
+    #                 keys will be started.
+    def initialize(token_string, shared_secret, merchant_id, logger: Logger.new($stdout), signing_keys: nil)
+      Aliquot.start_key_updater(logger) if $key_updater_thread.nil? && signing_keys.nil?
+
+      validation = Aliquot::Validator::Token.new(JSON.parse(token_string))
+      validation.validate
+
+      @token = validation.output
+
+      @shared_secret = shared_secret
+      @merchant_id   = merchant_id
+      @signing_keys  = signing_keys
+    end
+
+    ##
+    # Validate and decrypt the token.
+    def process
+      unless valid_protocol_version?
+        raise Error, 'only ECv1 protocolVersion is supported'
+      end
+
+      unless valid_signature?
+        raise InvalidSignatureError
+      end
+
+      @signed_message = JSON.parse(@token['signedMessage'])
+      validate(Aliquot::Validator::SignedMessage, @signed_message)
+
+      aes_key, mac_key = derive_keys(@signed_message['ephemeralPublicKey'],
+                                     @shared_secret,
+                                     'Google')
+
+      raise InvalidMacError unless valid_mac?(mac_key,
+                                              @signed_message['encryptedMessage'],
+                                              @signed_message['tag'])
+
+      @message = decrypt(aes_key, @signed_message['encryptedMessage'])
+
+      validate(Aliquot::Validator::EncryptedMessageValidator, @message)
+
+      raise ExpiredException if expired?
+
+      @message
+    end
+
+    def protocol_version
+      @token['protocolVersion']
+    end
+
+    def valid_protocol_version?
+      protocol_version == 'ECv1'
+    end
+
+    ##
+    # Check if the token is expired, according to the messageExpiration included
+    # in the token.
+    def expired?
+      @message['messageExpiration'].to_f / 1000.0 <= Time.now.to_f
+    end
+
+    def valid_signature?
+      signed_string = ['Google', @merchant_id, protocol_version, @token['signedMessage']].map do |str|
+        [str.length].pack('V') + str
+      end.join
+
+      keys = JSON.parse(signing_keys)['keys']
+      # Check if signature was performed with any possible key.
+      keys.map do |e|
```

Comment:
`e` is for entry.

It does need to, by the design of Google Pay.
It could be extracted to a function, but I really don't see any benefit to it.
Thought I'd be happy to do it.